File: 0680_validPalindromeII/main.cpp
#include <string>

class Solution {
public:
    bool validPalindrome(std::string s) {
        int l = 0;
        int r = s.size() - 1;
        while(l < r) {
            if(s[l] != s[r]) {
                int l1 = l, r1 = r-1, l2 = l+1, r2 = r;
                while(l1 < r1 && s[l1] == s[r1]) {
                    ++l1;
                    --r1;
                }

                while(l2 < r2 && s[l2] == s[r2]) {
                    ++l2;
                    --r2;
                }
                return (l1 >= r1) || (l2 >= r2);
            }
            ++l;
            --r;
        }
        return true;
    }

};

class Solution1 {
private:
    bool ispalindrome(std::string& s, int l, int r){
        while(l < r){
            if(s.at(l) == s.at(r)){
                ++l;
                --r;
            }else 
                return false;
        }
        return true;
    }
public:
    bool validPalindrome(std::string s) {
        int l  = 0;
        int r = s.size()-1;
        while(l < r){
            if(s.at(l) == s.at(r)){
                ++l;
                --r;
            }else
                return ispalindrome(s, l+1, r) || ispalindrome(s, l, r-1);
        }
        return true;
    }
};

File: 0680_validPalindromeII/main.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        def verify(l, r, deleted):
            while l < r:
                if s[l] != s[r]:
                    if deleted == 0:
                        return False
                    else:
                        return verify(l+1, r, deleted-1) or verify(l, r-1, deleted-1)
                l += 1
                r -= 1
            return True
        
        return verify(0, len(s)-1, 1)
    
    def validPalindrome1(self, s):
        i = 0
        while i < len(s) / 2 and s[i] == s[-(i + 1)]: 
            i += 1
            
        s = s[i:len(s) - i]
        return s[1:] == s[1:][::-1] or s[:-1] == s[:-1][::-1]
